casper/src/main/scala/coop/rchain/casper/SynchronyConstraintChecker.scala

```diff
@@ -35,40 +36,53 @@ final class SynchronyConstraintChecker[F[_]: Sync: BlockStore: Log](
-          synchronyConstraintValue = if (otherValidatorsWeight == 0) 1
-          else
-            sendersWeight.toDouble / otherValidatorsWeight
-          _ <- Log[F].info(
-                s"Seen ${seenSenders.size} senders with weight $sendersWeight out of total $otherValidatorsWeight " +
-                  s"(${synchronyConstraintValue} out of $synchronyConstraintThreshold needed)"
-              )
-        } yield synchronyConstraintValue >= synchronyConstraintThreshold
+          lastProposedBlockMeta <- dag.lookupUnsafe(lastProposedBlockHash)
+
+          checkConstraint = for {
+            // Guaranteed to be present since last proposed block was present
+            seenSenders <- calculateSeenSendersSince(lastProposedBlockMeta, dag)
+
+            // Loading the whole block is only needed to get post-state hash
+            lastProposedBlock     <- BlockStore[F].getUnsafe(lastProposedBlockHash)
+            lastProposedStateHash = ProtoUtil.postStateHash(lastProposedBlock)
+            // NOTE: It would be useful to have active validators cached in block in the same way as bonds.
+            activeValidators <- runtimeManager.getActiveValidators(lastProposedStateHash)
+
+            // Validators weight map filtered by active validators only.
+            validatorWeightMap = lastProposedBlockMeta.weightMap.filter {
+              case (validator, _) => activeValidators.contains(validator)
+            }
+            sendersWeight = seenSenders.toList.flatMap(validatorWeightMap.get).sum
+
+            // This method can be called on readonly node or not active validator.
+            // So map validator -> stake might not have key associated with the node,
+            // that's why we need `getOrElse`
+            otherValidatorsWeight = validatorWeightMap.values.sum - validatorWeightMap
+              .getOrElse(validator, 0L)
+            // If there is no other active validators, do not put any constraint (value = 1)
+            synchronyConstraintValue = if (otherValidatorsWeight == 0) 1
+            else
+              sendersWeight.toDouble / otherValidatorsWeight
+            _ <- Log[F].info(
+                  s"Seen ${seenSenders.size} senders with weight $sendersWeight out of total $otherValidatorsWeight " +
+                    s"(${synchronyConstraintValue} out of $synchronyConstraintThreshold needed)"
+                )
+          } yield synchronyConstraintValue >= synchronyConstraintThreshold
+
+          // Calculate if validator has latest proposed block older than approved block.
+          // NOTE: When LFS is restored, we don't have tuple space older then approved block,
+          //  so we cannot read active validators from PoS.
+          approvedBlockNumber       = ProtoUtil.blockNumber(approvedBlock)
+          latestBlockBehindApproved = approvedBlockNumber > lastProposedBlockMeta.blockNum
+
+          allowedToPropose <- if (latestBlockBehindApproved) true.pure[F] else checkConstraint
```

Comment:
Maybe it would be better to just skip active validators filtering instead of allow without checking. @nzpr can we do this or this will bring the old problem again resolved in PR #3069?

Reply:
@tgrospic I feel that using lastProposedTuplespace for fetching active validators is not correct - you have to use state that you are trying to propose on top of, so, basically, state of the main parent. Wdy think?

Reply:
@nzpr makes sense to use main parent. Do you know why is state from latest block used now?

Reply:
@tgrospic not really, I don't see why its more preferable then main parent.